packages/extension-commerce-bm-seo/src/setup-app.tsx

```diff
@@ -0,0 +1,172 @@
+     * This method is used to make changes to the PWA-Kit application routes. If your extension adds a new page to the application
+     * then you can add it to the router here. The routes passed to this method is an accrued list of routes that have been added
+     * from extensions applied before it. It is called during the `getRoutes` phase on both the server and the client.
+     *
+     * NOTE: If you instead want to modify a list of all the routes, refer to the `beforeRouteMatch` below.
+     */
+    async getRoutesAsync({locals}: GetRoutesParams): Promise<AsyncRouteProps[]> {
+        let urlMapping
+
+        // TODO: do we need this check?
+        if (!locals.originalUrl) {
+            return Promise.resolve([])
+        }
+
+        // Make SEO GET Url Mapping API call
+        const urlSegment: string = locals.originalUrl.split('?')[0]
+        const config = this.getConfig()
+        const shopperSeo = await getShopperSeoClient(locals, config)
+        try {
+            urlMapping = await shopperSeo.getUrlMapping({
+                parameters: {urlSegment}
+            })
+        } catch (e) {
+            console.error(`Couldn't find mapping for given segement: ${urlSegment}`)
+        }
+
+        if (!urlMapping) {
+            return Promise.resolve([])
+        }
+
+        const {component, props} = getComponentForUrlMapping(
+            urlMapping,
+            config.resourceTypeToComponentMap
+        )
+        const requestURL = new URL(urlSegment, getAppOrigin(locals))
+        return Promise.resolve([
+            {
+                path: requestURL.pathname,
+                component,
+                componentProps: props,
+                exact: true
+            }
+        ])
+    }
+
+    /**
+     * TODO: update comment to make it clearer that beforeRouteMatch is also called on the client side
+     * This method is used on the server during the rendering pipeline. It's provided a list of all the routes that your application
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({allRoutes}: BeforeRouteMatchParams): RouteProps[] {
+        const updatedRoutes = [...allRoutes]
+        const filteredRoutes = updatedRoutes.filter((route) => !route.component?.isPlaceholder)
+
+        for (let i = 0; i < updatedRoutes.length; i++) {
+            const route = updatedRoutes[i]
+            if (!route.component.isPlaceholder) continue
+
+            const componentName: string = route.component.displayName
+            const actualComponent = filteredRoutes.find((r) =>
+                r.component?.displayName?.includes(componentName)
+            )?.component
+
+            if (!actualComponent) {
+                throw new Error(`Could not find component with displayName "${componentName}"`)
```

Comment:
Minor: because `displayName` can be longer and does not match 1 to 1 with `componentName`:

```suggestion
                throw new Error(`Could not find component with displayName containing "${componentName}"`)
```